src/appleseed/renderer/modeling/frame/frame.cpp

```diff
@@ -462,28 +490,465 @@ void Frame::denoise(
+        }
+    }
+
+    RENDERER_LOG_INFO("opened checkpoint resuming at pass %s",
+        pretty_uint(start_pass).c_str());
+
+    impl->m_initial_pass = start_pass;
+
+    return true;
+}
+
+void Frame::save_checkpoint(
+    PermanentShadingResultFrameBufferFactory*   buffer_factory,
+    const size_t                                pass) const
+{
+    if (!impl->m_checkpoint_create)
+        return;
+
+    create_parent_directories(impl->m_checkpoint_create_path.c_str());
+
+    GenericImageFileWriter writer(impl->m_checkpoint_create_path.c_str());
+
+    ImageAttributes image_attributes = ImageAttributes::create_default_attributes();
+    image_attributes.insert("appleseed:LastPass", pass);
+
+    // Create layers.
+    // Buffer containing pixels' weight.
+    ShadingBufferCanvas weight_canvas(*this, buffer_factory);
+
+    // Create channel names.
+    const size_t shading_channel_count = weight_canvas.properties().m_channel_count;
+    vector<string> shading_channel_names; // for shading buffer
+    vector<const char *> shading_channel_names_cr;
+
+    static const string channel_name_prefix = "channel_";
+    for (size_t i = 0; i < shading_channel_count; ++i)
+    {
+        shading_channel_names.push_back(channel_name_prefix + pad_left(to_string(i + 1), '0', 4));
+        shading_channel_names_cr.push_back(shading_channel_names[i].c_str());
+    }
+
+    assert(shading_channel_names.size() == shading_channel_count);
+
+    // Add layers in the file.
+    writer.append_image(&image());
+    image_attributes.insert("image_name", "beauty");
+    writer.set_image_attributes(image_attributes);
+
+    writer.append_image(&weight_canvas);
+    image_attributes.insert("image_name", "appleseed:RenderingBuffer");
+    writer.set_image_channels(shading_channel_count, shading_channel_names_cr.data());
+    writer.set_image_attributes(image_attributes);
+
+    // Add AOV layers.
+    for (size_t i = 0, e = aovs().size(); i < e; ++i)
+    {
+        const AOV* aov = aovs().get_by_index(i);
+
+        const char* aov_name = aov->get_name();
+        const Image& aov_image = aov->get_image();
+        writer.append_image(&aov_image);
+        image_attributes.insert("image_name", aov_name);
+        writer.set_image_channels(aov->get_channel_count(), aov->get_channel_names());
+        writer.set_image_attributes(image_attributes);
+    }
+
+    // Write the file.
+    writer.write();
+
+    // Add internal AOVs layers (in external files).
+    for (size_t i = 0, e = internal_aovs().size(); i < e; ++i)
+    {
+        const AOV* aov = internal_aovs().get_by_index(i);
+
+        const DenoiserAOV* denoiser_aov = dynamic_cast<const DenoiserAOV*>(aov);
+
+        // Save denoiser checkpoint.
+        if (denoiser_aov != nullptr)
+        {
+            save_denoiser_checkpoint(impl->m_checkpoint_create_path, denoiser_aov);
+        }
+    }
+
+    RENDERER_LOG_INFO("wrote checkpoint for pass %s.", pretty_uint(pass + 1).c_str());
```

Comment:
Let's mention the checkpoint file path.

    wrote checkpoint file %s for pass %s.